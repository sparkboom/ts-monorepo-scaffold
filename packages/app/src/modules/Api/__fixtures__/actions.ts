export const persistRehydrateWithToken = {
  type: 'persist/REHYDRATE',
  payload: {
    api: {
      token: {
        jwt: 'PERSISTED_TOKEN',
        expiryTimeStamp: 1571615706078,
        expiresInSeconds: 3311.836,
      },
    },
    _persist: {
      version: -1,
      rehydrated: true,
    },
  },
  key: '%%APP_NAME%%',
};

export const persistRehydrateNoToken = {
  type: 'persist/REHYDRATE',
  payload: {
    api: {
      token: {},
    },
    _persist: {
      version: -1,
      rehydrated: true,
    },
  },
  key: '%%APP_NAME%%',
};
