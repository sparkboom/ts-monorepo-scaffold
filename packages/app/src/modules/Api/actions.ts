import { Action } from 'redux';

// Types
interface ApiSetTokenAction {
  (jwt: string, expiryTimeStamp: number): AppAction<'API__SET_TOKEN', { jwt: string; expiryTimeStamp: number }>;
}
interface ApiInitAction {
  (): Action<'API__INIT'>;
}

// Actions
export const apiSetToken: ApiSetTokenAction = (jwt, expiryTimeStamp) => ({
  type: 'API__SET_TOKEN',
  payload: {
    jwt,
    expiryTimeStamp,
  },
});

export const apiInit: ApiInitAction = () => ({
  type: 'API__INIT',
});
