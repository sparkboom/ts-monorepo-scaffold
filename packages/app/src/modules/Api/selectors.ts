import { createSelector } from 'reselect';
import { ClientCmdApi, ClientQueryApi, Configuration, ClientWebApi } from '@%%APP_NAME%%/api';
import { get } from 'lodash-es';
import { queryEndpointSelector, commandEndpointSelector, webEndpointSelector } from '../config/selectors';
import { Endpoint } from '../config/types';
import { Api, Token } from './types';
import { Search } from 'history';

// Types
type ApiClient = typeof ClientCmdApi | typeof ClientQueryApi | typeof ClientWebApi;
interface CreateApi {
  (Api: ApiClient, endpoint: Endpoint | null, token: Api['token']): ClientCmdApi | ClientQueryApi | ClientWebApi | null;
}

// Helpers
const createApi: CreateApi = (Api, endpoint, token) => {
  if (!endpoint || !token || !token.jwt) {
    return null;
  }
  const apiConfig = new Configuration({
    basePath: endpoint.baseUrl,
    accessToken: token.jwt,
  });
  return new Api(apiConfig);
};

// Selectors
export const routerQuerySelector = (state: RootState): Search | null => get(state, 'router.location.search', null);
export const tokenSelector = (state: RootState): Token | null => {
  const rawToken: Token | null = get(state, 'api.token', null);
  if (!rawToken) {
    return null;
  }
  return {
    ...rawToken,
    expiryTimeStamp: Number(rawToken.expiryTimeStamp),
  };
};

// Memoized Selectors
export const querySelector = createSelector(
  routerQuerySelector,
  (routerQuery: Search | null) => {
    if (!routerQuery) {
      return {};
    }
    const kvPairs = routerQuery.substr(1).split('&');
    const queryByKey = kvPairs.reduce((acc, q) => {
      const kv = q.split('=');
      return {
        ...acc,
        [kv[0]]: kv[1],
      };
    }, {});
    return queryByKey;
  },
);

export const queryApiSelector = createSelector(
  queryEndpointSelector,
  tokenSelector,
  (queryEndpoint, token) => {
    const api = createApi(ClientQueryApi, queryEndpoint, token) as ClientQueryApi;
    return api;
  },
);

export const commandApiSelector = createSelector(
  commandEndpointSelector,
  tokenSelector,
  (commandEndpoint, token) => {
    const api = createApi(ClientCmdApi, commandEndpoint, token) as ClientCmdApi;
    return api;
  },
);

export const webApiSelector = createSelector(
  webEndpointSelector,
  webEndpoint => {
    if (!webEndpoint) {
      return null;
    }
    const apiConfig = new Configuration({
      basePath: `${webEndpoint.baseUrl}${webEndpoint.basePath}`,
    });
    return new ClientWebApi(apiConfig);
  },
);
