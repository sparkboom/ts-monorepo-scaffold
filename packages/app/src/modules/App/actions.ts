import { CheckStatus, JwtError, ConfigError, Check } from './types';

// Types
interface AppInitAction {
  (): {
    type: 'APP__INIT';
    correlationId: '%%APP_NAME%%_KERNEL';
  };
}
interface AppSetCheckStatusAction {
  (check: Check, status: CheckStatus, error: JwtError | ConfigError | null): {
    type: 'APP__SET_CHECK_STATUS';
    payload: {
      check: Check;
      status: CheckStatus;
      error: string | null;
    };
  };
}

// Actions
export const appInit: AppInitAction = () => ({
  type: 'APP__INIT',
  correlationId: '%%APP_NAME%%_KERNEL',
});

export const appSetCheckStatus: AppSetCheckStatusAction = (check, status, error = null) => ({
  type: 'APP__SET_CHECK_STATUS',
  payload: {
    check,
    status,
    error,
  },
});
