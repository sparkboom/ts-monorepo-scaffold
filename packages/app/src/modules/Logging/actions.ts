/* eslint-disable @typescript-eslint/no-explicit-any */
import { Action } from 'redux';
import { get } from 'lodash-es';
import { LogLevel } from './types';

// Types
type Details = any | null;
type ResponseType = 'STATUS' | 'FAIL' | 'WARN' | 'SUCCESS';
interface IssueAction<T extends string> {
  type: string;
  payload: {
    message: string | null;
    details: Details | null;
    relatedAction: Action<T>;
  };
  correlationId: string;
}
export interface SubmitLogAction {
  type: 'LOGGING__SUBMIT';
  payload: {
    level: LogLevel;
    context: string;
    message: string;
    date: Date;
    details: any[];
  };
}

export interface SubmitErrorAction {
  type: 'LOGGING__ERROR_SUBMIT';
  payload: {
    level: LogLevel;
    context: string;
    message: string | null;
    error: Error;
    date: Date;
    details: any[];
  };
}

// Sagas
const response = <T extends string>(type: ResponseType, relatedAction: Action<T>, message: string | null = null, details: Details = null): IssueAction<T> => ({
  type: `${relatedAction.type}_${type}`,
  payload: {
    message,
    details,
    relatedAction,
  },
  correlationId: `${get(relatedAction, 'correlationId')}`,
});

export const fail = <T extends string>(relatedAction: Action<T>, message = '', details: Details = null): IssueAction<T> =>
  response('FAIL', relatedAction, message, details);

export const warn = <T extends string>(relatedAction: Action<T>, message = '', details: Details = null): IssueAction<T> =>
  response('WARN', relatedAction, message, details);

export const success = <T extends string>(relatedAction: Action<T>, details: Details = null): IssueAction<T> =>
  response('SUCCESS', relatedAction, null, details);

export const status = <T extends string>(relatedAction: Action<T>, statusValue: string): IssueAction<T> =>
  response('STATUS', relatedAction, null, { status: statusValue });

export const submitError = (level: LogLevel, context: string, messageError: Error | string | null, ...details: any[]): SubmitErrorAction => {
  let error = null;
  let message = null;
  if (messageError && messageError instanceof Error) {
    error = messageError;
    message = messageError.message || 'Unknown error';
  } else {
    message = messageError || 'Unknown error';
    // We artificially create an Error object so that we can construct a stack trace later
    // This technique is supported by all browsers including IE11
    try {
      throw Error();
    } catch (err) {
      error = err;
    }
  }
  return {
    type: 'LOGGING__ERROR_SUBMIT',
    payload: {
      level,
      context,
      message,
      error,
      date: new Date(),
      details,
    },
  };
};

export const submitLog = (level: LogLevel, context: string, message: string, ...details: any[]): SubmitLogAction => ({
  type: 'LOGGING__SUBMIT',
  payload: {
    level,
    context,
    message,
    date: new Date(),
    details,
  },
});
