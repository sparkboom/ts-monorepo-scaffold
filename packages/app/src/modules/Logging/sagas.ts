/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/unbound-method */
import StackTrace from 'stacktrace-js';
import { call, SimpleEffect, CallEffectDescriptor, takeEvery, select, apply, SelectEffectDescriptor } from 'redux-saga/effects';
import { ClientCmdApi } from '@adstudio-leadcap/api';
import { SubmitLogAction, SubmitErrorAction } from './actions';
import { commandApiSelector } from '../Api/selectors';
import { LogLevel } from './types';

// Constants
const LOG_MAP = {
  NO_LOG: 1,
  ERROR: 2,
  WARN: 3,
  INFO: 4,
  DEBUG: 5,
};
const BROWSER_LOG_LEVEL = LOG_MAP.INFO;
const SERVICE_LOG_LEVEL = LOG_MAP.WARN;
const APP_CONTACT = 'LEADCAP';
const INITIAL_TIMESTAMP = new Date().getTime();
const MAX_STACK_TRACE_LENGTH = 4000;

// Types
type CallEffect = SimpleEffect<'CALL', CallEffectDescriptor<any>>;
type SelectEffect = SimpleEffect<'Select', SelectEffectDescriptor>;
type LogSaga = Generator<CallEffect | SelectEffect, void, StackTrace.StackFrame[] & ClientCmdApi>;
interface TimeStamp {
  longUtc: string;
  shortLocal: string;
  msSinceLaunch: string;
}
interface BrowserLogger {
  (
    timeStamp: TimeStamp,
    level: LogLevel,
    context: string,
    message: string | null,
    stackTrace: StackTrace.StackFrame[] | null,
    details: any[],
    logger?: ConsoleLogger,
  ): void;
}
interface ServiceLogger {
  (timeStamp: TimeStamp, level: LogLevel, context: string, message: string | null, stackTrace: StackTrace.StackFrame[] | null, details: any[]): Generator<
    any,
    any,
    any
  >;
}

/**
 * formatNumber
 *
 * Formats numbers using english/american format using a comma as dividing character.
 * Useful when formatting without locale requirements, such as logging.
 */
function formatNumber(num: number): string {
  return num.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');
}

/**
 * getTimeStamp
 *
 * Converts a passed date to various formats for logging
 * - long UTC is useful for logging to a service
 * - shortLocal is useful for logging to local browser console
 * - msSinceLaunch is useful for monitoring performance and sorting log entries
 */
export function getTimeStamp(date: Date = new Date()): TimeStamp {
  const dateStr = {
    longUtc: '',
    shortLocal: '',
    msSinceLaunch: '',
  };
  try {
    dateStr.longUtc = `${date.getUTCFullYear()}-${date.getUTCMonth()}-${date.getUTCDate()} ${date.getUTCHours()}:${date.getUTCMinutes()}:${date.getUTCSeconds()}.${date.getUTCMilliseconds()}`;
    dateStr.shortLocal = `${date
      .getFullYear()
      .toString()
      .substring(2)}-${date.getMonth()}-${date.getDate()} ${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}.${date.getMilliseconds()}`;
    dateStr.msSinceLaunch = `+${formatNumber(new Date().getTime() - INITIAL_TIMESTAMP)} ms`;
  } catch (err) {
    console.error('Could not format current date', date);
  }
  return dateStr;
}

/**
 * logToConsole
 *
 * Formats all arguments consistently to produce a console log, and provide a collapsed content to display further details
 */
export const logToConsole: BrowserLogger = (timeStamp, level, context, message, stackTrace, details, logger = console) => {
  const log = {
    ERROR: logger.error,
    WARN: logger.warn,
    INFO: logger.info,
    DEBUG: logger.debug,
  }[level];

  log(`${timeStamp.shortLocal} ${APP_CONTACT}.${context}: ${message}`, details, stackTrace);
};

/**
 * logToService
 *
 * Formats all arguments consistently to produce a console log, and provide a collapsed content to display further details
 */
// eslint-disable-next-line func-names
export const logToServiceSaga: ServiceLogger = function*(timeStamp, level, context, message, stackTrace, details) {
  const cmdApi = yield select(commandApiSelector);
  if (!cmdApi) {
    return;
  }
  try {
    const payload = {
      body: {
        message,
        context,
        level,
        timeStamp: timeStamp.longUtc,
        stackTrace: `${stackTrace}`.substring(0, MAX_STACK_TRACE_LENGTH),
        details,
      },
    };
    yield apply(cmdApi, cmdApi.saveFrontEndLog, [payload]);
  } catch (err) {
    // Fallback
    console.error(err);
  }
};

// Sagas

/**
 * loggingSaga
 *
 * A worker saga that will log to the console if the log level matches the BROWSER_LOG_LEVEL or higher
 * and also submits the log to the frontEndLog endpoint if the level matches the LOGGER_LOG_LEVEL or higher
 */
export function* logSaga(action: SubmitLogAction | SubmitErrorAction, browserLogLevel = BROWSER_LOG_LEVEL, serviceLogLevel = SERVICE_LOG_LEVEL): LogSaga {
  try {
    let stackTrace: StackTrace.StackFrame[] | null = null;
    if (action.type === 'LOGGING__ERROR_SUBMIT') {
      stackTrace = yield call([StackTrace, StackTrace.fromError], action.payload.error);
    }
    const { level, message, context, date, details } = action.payload;

    const logLevelScalar = LOG_MAP[level];
    const ts = yield call(getTimeStamp, date);

    if (logLevelScalar <= browserLogLevel) {
      yield call(logToConsole, ts, level, context, message, stackTrace, details);
    }

    if (logLevelScalar <= serviceLogLevel) {
      yield call(logToServiceSaga, ts, level, context, message, stackTrace, details);
    }
  } catch (err) {
    console.error(err);
  }
}

/**
 * loggingSaga
 *
 * The root saga for the Logging module
 */
export default function* loggingSaga() {
  yield takeEvery(['LOGGING__SUBMIT', 'LOGGING__ERROR_SUBMIT'], logSaga);
}
