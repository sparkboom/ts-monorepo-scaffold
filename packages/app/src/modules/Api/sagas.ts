import {
  call,
  delay,
  apply,
  select,
  take,
  SimpleEffect,
  TakeEffectDescriptor,
  PutEffectDescriptor,
  SelectEffectDescriptor,
  CallEffectDescriptor,
  takeLatest,
  ForkEffectDescriptor,
} from 'redux-saga/effects';
import { Action } from 'redux';
import { get } from 'lodash-es';
import { ClientCmdApi, ClientQueryApi, ClientWebApi } from '@%%APP_NAME%%/api';
import { apiSetToken } from './actions';
import { querySelector, commandApiSelector, queryApiSelector, webApiSelector } from './selectors';
import { putContext } from '../Logging/effects';
import { NOTIFICATION__XTRA_LONG_EXPIRY_MS, NOTIFICATION__LONG_EXPIRY_MS } from '../Notification/actions';
import { DetailedError } from '../../common/detailedError';
import { isToken } from './types';
import { notifySaga } from '../Notification/sagas';

// Constants
export const LOG_CTX = 'API.SAGAS';
export const MIN_QUEUE_MSECS = 10000;
export const JWT_EXPIRY_BUFFER_SECS = 300;
export const MS_IN_SECONDS = 1000;
export const OBFUSCATE_TOKEN_VISIBLE_CHARS = 8;
export const ERR__TOKEN_REFRESH_TERMINATED = 'Token refresh cycle permenantly terminated.';

// Types
interface Token {
  jwt: string;
  expiryTimeStamp: number;
  expiresInSeconds: number;
}
interface TokenFromApi {
  authToken: string;
  expiresInSeconds: string;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type GeneratorNextAny = any;
type TakeEffect = SimpleEffect<'TAKE', TakeEffectDescriptor>;
type PutEffect = SimpleEffect<'PUT', PutEffectDescriptor<any>>;
type CallEffect = SimpleEffect<'CALL', CallEffectDescriptor<any>>;
type SelectEffect = SimpleEffect<'SELECT', SelectEffectDescriptor>;
type ForkEffect = SimpleEffect<'FORK', ForkEffectDescriptor<never>>;
type ApiInitSaga = Generator<TakeEffect | PutEffect | CallEffect | SelectEffect, void, GeneratorNextAny>;
type ApiSaga = Generator<ForkEffect, void, unknown>;
type RefreshTokenSaga = Generator<SelectEffect | CallEffect | PutEffect, Token | null, GeneratorNextAny>;
type GetQueryApiSaga = Generator<SelectEffect | PutEffect, ClientQueryApi | null, GeneratorNextAny>;
type GetCommandApiSaga = Generator<SelectEffect | PutEffect, ClientCmdApi | null, GeneratorNextAny>;
type GetWebApiSaga = Generator<SelectEffect | PutEffect, ClientWebApi | null, GeneratorNextAny>;
type GetSessionTokenSaga = Generator<PutEffect, Token | null, unknown>;

// Helpers
export function createToken(jwt: string | null, expiresInSeconds: string | number, date: Date | undefined = new Date()): Token | null {
  if (!jwt) {
    return null;
  }
  const expiresInSecondsNum = Number(expiresInSeconds);
  const expiryDate = date;
  expiryDate.setSeconds(expiryDate.getSeconds() + (expiresInSecondsNum || 0));
  return {
    jwt,
    expiryTimeStamp: expiryDate.getTime(),
    expiresInSeconds: expiresInSecondsNum,
  };
}

/* we're passing optional DateTime so we can easily test the function */
/**
 * getSecondsRemaining calculates the time remaining in seconds, of the timestamp provided.
 *
 * @param expiryTimeStamp - represents the number of milliseconds since the Unix epoch. This value can represent the past, present or future.
 * @param DateTime - the Date class. This parameter promotes easier unit testing.
 * @returns total seconds difference between now and the expiryTimeStamp, a negative value represents reflects a timestamp in the past,
 * and a positive value representing a future time.
 */
export function getSecondsRemaining(expiryTimeStamp: number, DateTime = Date): number {
  const nowTimeStamp = new DateTime().getTime();
  const deltaMilliseconds = expiryTimeStamp - nowTimeStamp;
  return deltaMilliseconds / MS_IN_SECONDS;
}

// obfuscateToken is used for logging
function obfuscateToken(token: Token | null): { obfuscatedToken: Token; summary: string } | null {
  if (!token) {
    return null;
  }
  const obfuscatedToken = { ...token };
  const hiddenCharLength = Math.max(2, obfuscatedToken.jwt.length - OBFUSCATE_TOKEN_VISIBLE_CHARS);
  const visibleTokenChars = obfuscatedToken.jwt.substr(hiddenCharLength);
  obfuscatedToken.jwt = '*'.repeat(hiddenCharLength) + visibleTokenChars;
  const summary = `token=*${visibleTokenChars}, ts=${obfuscatedToken.expiryTimeStamp}, expires=${obfuscatedToken.expiresInSeconds}`;
  return {
    obfuscatedToken,
    summary,
  };
}
export const refreshTokenDelayMsFromExpiry = (expiresInSeconds: number): number =>
  Math.max(MIN_QUEUE_MSECS, (expiresInSeconds - JWT_EXPIRY_BUFFER_SECS) * MS_IN_SECONDS);

// Sagas
export function* refreshTokenSaga(action: Action<string>): RefreshTokenSaga {
  const put = putContext(action, LOG_CTX);
  try {
    yield put.logDebug(`Token refresh start`);

    // 1. Get Cmd API
    const cmdApi = (yield select(commandApiSelector)) as ClientCmdApi | null;
    if (!cmdApi) {
      throw new DetailedError('Token Refresh Failed. Command API unavailable', { cmdApi });
    }

    // 2. Refresh Token
    // eslint-disable-next-line @typescript-eslint/unbound-method
    const rawToken: TokenFromApi | null = yield apply(cmdApi, cmdApi.authenticationRefreshToken, []);
    if (!rawToken) {
      throw new DetailedError('Token Refresh Failed. No valid token received from API refresh call.', { rawToken });
    }

    // 3. Transform Token
    const newToken: Token | null = yield call(createToken, rawToken.authToken, rawToken.expiresInSeconds);
    if (!newToken || !newToken.jwt) {
      throw new DetailedError('Token Refresh Failed. Token is possibly invalid.', { rawToken, newToken });
    }

    // 4. Store New Token
    yield put(apiSetToken(newToken.jwt, newToken.expiryTimeStamp));
    const obfuscatedToken = obfuscateToken(newToken);
    yield put.logInfo(`API Token successfully updated. ${obfuscatedToken && obfuscatedToken.summary}`, {
      token: obfuscatedToken && obfuscatedToken.obfuscatedToken,
    });
    if (sessionStorage) {
      sessionStorage.setItem('jwt', newToken.jwt);
      sessionStorage.setItem('expires', `${newToken.expiryTimeStamp}`);
    }

    return newToken;
  } catch (error) {
    yield put.errorWarn(error);
    yield call(
      notifySaga,
      action,
      'common.notifications.sessionRefreshFailed.heading',
      'common.notifications.sessionRefreshFailed.details',
      'ERROR',
      NOTIFICATION__XTRA_LONG_EXPIRY_MS,
    );
  }

  return null;
}

/**
 * getSessionTokenSaga obtains the available tokens from session storage, one from V1 FE app, the other from V2 FE app
 * It will compare the two and use the one that expires last, thus assuming it is the most recently retrieved token.
 *
 * When V1 FE has been retired, then we can remove this complication, and simply use 'yield select(tokenSelector)'
 *
 * @param action - triggering action for the saga
 * @param currentV2Token - we already have this value, so we'll pass it instead of retrieving it
 * @param global - typically represents the global window object. Providing the value as an argument allows this generator to be tested
 */
export function* getSessionTokenSaga(action: Action<string>, currentV2Token: Token | null, global: Window & typeof globalThis = window): GetSessionTokenSaga {
  const put = putContext(action, LOG_CTX);
  let preferredToken: Token | null = null;

  // 1. Construct V2 Token via redux-persist
  if (currentV2Token) {
    // eslint-disable-next-line no-param-reassign
    currentV2Token.expiresInSeconds = getSecondsRemaining(currentV2Token.expiryTimeStamp);
  }

  // 2. Get V1 Token via session storage if available and valid
  let currentV1Token: Token | null = null;
  const { sessionStorage } = global;
  if (sessionStorage) {
    const jwtV1 = sessionStorage.getItem('jwt');
    const expiresV1 = Number(sessionStorage.getItem('expires'));
    if (jwtV1 && expiresV1 && !Number.isNaN(expiresV1)) {
      currentV1Token = {
        jwt: jwtV1,
        expiryTimeStamp: expiresV1,
        expiresInSeconds: getSecondsRemaining(expiresV1),
      };
    }
  }

  // 3. Choose preferred token with the latter expiry
  if (!currentV1Token && currentV2Token) {
    preferredToken = currentV2Token;
  } else if (!currentV2Token && currentV1Token) {
    preferredToken = currentV1Token;
  } else if (currentV2Token && currentV1Token) {
    preferredToken = currentV1Token.expiryTimeStamp > currentV2Token.expiryTimeStamp ? currentV1Token : currentV2Token;
  } else {
    return null;
  }

  // 4. Ensure both V1 and V2 formats are the preferred token
  yield put(apiSetToken(preferredToken.jwt, preferredToken.expiryTimeStamp));
  if (sessionStorage) {
    sessionStorage.setItem('jwt', preferredToken.jwt);
    sessionStorage.setItem('expires', `${preferredToken.expiryTimeStamp}`);
  }

  return preferredToken;
}

export function* apiInitSaga(action: Action<string>, global: Window & typeof globalThis = window): ApiInitSaga {
  const put = putContext(action, LOG_CTX);
  try {
    // 1. Ensure redux-persist has rehydrated
    const persistState = yield take('persist/REHYDRATE');

    // 2. Set JWT app check to in-progress
    yield put.status('IN_PROGRESS');

    // 3. Choose token
    const searchQueries: Record<string, string> = yield select(querySelector);
    const queryToken = searchQueries ? createToken(searchQueries.jwt, searchQueries.expiresInSeconds) : null;
    let token: Token | null = null;

    debugger;

    if (isToken(queryToken)) {
      // 3a. Load token from query parameters
      // This token takes precedence, it happens on the initial redirect from the login endpoint

      // Store token in session storage in v1 & v2 format
      yield put(apiSetToken(queryToken.jwt, queryToken.expiryTimeStamp));
      if (sessionStorage) {
        sessionStorage.setItem('jwt', queryToken.jwt);
        sessionStorage.setItem('expires', `${queryToken.expiryTimeStamp}`);
      }

      // Remove query parameters, without pushing an entry in the browser navigation stack.
      const { history, document, location } = global;
      history.replaceState({}, document.title, location.origin + location.pathname);

      token = queryToken;
    } else {
      // 3b. Get most recent token from session storage - this could be v1 or v2 format
      // This happens when the user navigates using the back button, refreshes the frame, or redirects from the v1 app.
      const v2Token: Token | null = get(persistState, 'payload.api.token', null);
      token = yield* getSessionTokenSaga(action, v2Token);
    }

    // If there is no token in session storage, we cannot continue.
    if (!token) {
      throw new DetailedError('Session failed. No token found in querystring or session storage.');
    }

    // 4. Mark Token Check as success
    yield put.success();

    // 5. Token Refresh Cycle
    while (token !== null) {
      const refreshDelayMs = refreshTokenDelayMsFromExpiry(token.expiresInSeconds);
      yield put.logDebug(`Token refresh cycle started. Time until next refresh ${refreshDelayMs}ms`, {
        expiryTimeStamp: token.expiryTimeStamp,
        expiresInSeconds: token.expiresInSeconds,
      });

      yield delay(refreshDelayMs);
      token = (yield call(refreshTokenSaga, action)) as Token | null;
    }

    yield put.error(ERR__TOKEN_REFRESH_TERMINATED);
    yield call(
      notifySaga,
      action,
      'common.notifications.sessionTerminated.heading',
      'common.notifications.sessionTerminated.details',
      'ERROR',
      NOTIFICATION__LONG_EXPIRY_MS,
    );
  } catch (err) {
    yield put.error(err);
    yield call(
      notifySaga,
      action,
      'common.notifications.sessionFailed.heading',
      'common.notifications.sessionFailed.details',
      'ERROR',
      NOTIFICATION__XTRA_LONG_EXPIRY_MS,
    );
  } finally {
    yield put.fail();
  }
}

export function* getQueryApi(action: Action<any>): GetQueryApiSaga {
  const put = putContext(action, LOG_CTX);
  const queryApi: ClientQueryApi | null = yield select(queryApiSelector);

  if (!queryApi) {
    yield put.errorWarn('Query API could not be retrieved');
    return null;
  }

  return queryApi;
}

export function* getCommandApi(action: Action<any>): GetCommandApiSaga {
  const put = putContext(action, LOG_CTX);
  const commandApi: ClientCmdApi | null = yield select(commandApiSelector);

  if (!commandApi) {
    yield put.errorWarn('Command API could not be retrieved');
    return null;
  }

  return commandApi;
}

export function* getWebApi(action: Action<any>): GetWebApiSaga {
  const put = putContext(action, LOG_CTX);
  const webApi: ClientWebApi | null = yield select(webApiSelector);
  if (!webApi) {
    yield put.errorWarn('Web API could not be created');
    return null;
  }
  return webApi;
}

export default function* apiSaga(): ApiSaga {
  yield takeLatest('API__INIT', apiInitSaga);
}
