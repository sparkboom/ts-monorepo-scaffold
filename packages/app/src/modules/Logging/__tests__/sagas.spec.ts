import { expectSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import StackTrace from 'stacktrace-js';
import * as sagas from '../sagas';
import * as actions from '../actions';
import { SubmitErrorAction, SubmitLogAction } from '../actions';
import { commandApiSelector } from '../../Api/selectors';

jest.mock('stacktrace-js');

const mockedConsoleInfo = jest.spyOn(console, 'info');

// Tests
describe('Logging > Sagas', () => {
  let LOG_CTX: string;
  let FAKE_CMD_API: any;
  let FAKE_DETAILS: any;
  let FAKE_ERROR: Error;
  let FAKE_MESSAGE: string;
  let FAKE_ERROR_ACTION: SubmitLogAction | SubmitErrorAction;
  let FAKE_LOG_ACTION: SubmitLogAction | SubmitErrorAction;
  let FAKE_NULL_STACK_TRACE: null;
  let NO_LOG_LEVEL: 1;
  let ERROR_LOG_LEVEL: 2;

  beforeAll(() => {
    LOG_CTX = 'LOG.CTX';
    FAKE_CMD_API = {
      saveFrontEndLog: () => {},
    };
    FAKE_DETAILS = {
      details: 'Fake Details',
    };
    FAKE_ERROR = new Error('FAKE_ERROR_MSG');
    FAKE_MESSAGE = 'FAKE LOG MESSAGE';
    FAKE_ERROR_ACTION = actions.submitError('ERROR', LOG_CTX, FAKE_ERROR, FAKE_DETAILS);
    FAKE_LOG_ACTION = actions.submitLog('INFO', LOG_CTX, FAKE_MESSAGE, FAKE_DETAILS);
    FAKE_NULL_STACK_TRACE = null;
    NO_LOG_LEVEL = 1;
    ERROR_LOG_LEVEL = 2;

    mockedConsoleInfo.mockImplementation(() => {});
  });

  afterAll(() => {
    mockedConsoleInfo.mockReset();
  });

  describe('Log Saga', () => {
    it("should generate a stack trace if we're logging an error", () => {
      // Arrange
      // Act
      // Assert
      return expectSaga(sagas.logSaga, FAKE_ERROR_ACTION, NO_LOG_LEVEL, NO_LOG_LEVEL)
        .provide([
          [matchers.call.fn(StackTrace.fromError), FAKE_NULL_STACK_TRACE],
          [matchers.select(commandApiSelector), FAKE_CMD_API],
          [matchers.call.fn(FAKE_CMD_API.saveFrontEndLog), null],
        ])
        .call.fn(StackTrace.fromError)
        .run();
    });

    it("should not generate a stack trace if we're not logging an error", () => {
      // Arrange
      // Act
      // Assert
      return expectSaga(sagas.logSaga, FAKE_LOG_ACTION, NO_LOG_LEVEL, NO_LOG_LEVEL)
        .provide([
          [matchers.call.fn(StackTrace.fromError), FAKE_NULL_STACK_TRACE],
          [matchers.select(commandApiSelector), FAKE_CMD_API],
          [matchers.call.fn(FAKE_CMD_API.saveFrontEndLog), null],
        ])
        .not.call.fn(StackTrace.fromError)
        .run();
    });

    it('should not generate a stack trace if the error action does not contain an error object', () => {
      // Arrange
      const FAKE_ERROR_ACTION_NO_ERROR = {
        ...FAKE_ERROR_ACTION,
      };
      FAKE_ERROR_ACTION_NO_ERROR.payload.error = null;
      // Act
      // Assert
      return expectSaga(sagas.logSaga, FAKE_LOG_ACTION, NO_LOG_LEVEL, NO_LOG_LEVEL)
        .provide([[matchers.call.fn(StackTrace.fromError), FAKE_NULL_STACK_TRACE]])
        .not.call.fn(StackTrace.fromError)
        .run();
    });

    it('should log to browser if log level is same to browser log level', () => {
      // Arrange
      const INFO_LOG_LEVEL = 4;
      // Act
      // Assert
      return expectSaga(sagas.logSaga, FAKE_LOG_ACTION, INFO_LOG_LEVEL, NO_LOG_LEVEL)
        .provide([[matchers.call.fn(StackTrace.fromError), FAKE_NULL_STACK_TRACE]])
        .call.like({ fn: sagas.logToConsole })
        .run();
    });

    it('should log to browser if log level is above to browser log level', () => {
      // Arrange
      const DEBUG_LOG_LEVEL = 5;
      // Act
      // Assert
      return expectSaga(sagas.logSaga, FAKE_LOG_ACTION, DEBUG_LOG_LEVEL, NO_LOG_LEVEL)
        .provide([[matchers.call.fn(StackTrace.fromError), FAKE_NULL_STACK_TRACE]])
        .call.like({ fn: sagas.logToConsole })
        .run();
    });

    it('should not log to browser if log level is below to browser log level', () => {
      // Arrange
      const WARN_LOG_LEVEL = 3;
      // Act
      // Assert
      return expectSaga(sagas.logSaga, FAKE_LOG_ACTION, WARN_LOG_LEVEL, NO_LOG_LEVEL)
        .provide([[matchers.call.fn(StackTrace.fromError), FAKE_NULL_STACK_TRACE]])
        .not.call.like({ fn: sagas.logToConsole })
        .run();
    });

    it('should match the snapshot if content was successfully logged to browser', async () => {
      // Arrange
      const INFO_LOG_LEVEL = 4;
      const FAKE_TIMESTAMP = {
        longUtc: '2019-10-4 20:39:58.328',
        shortLocal: '19-10-4 15:39:58.328',
        msSinceLaunch: '+45 ms',
      };
      // Act
      // Assert
      const result = await expectSaga(sagas.logSaga, FAKE_LOG_ACTION, INFO_LOG_LEVEL, NO_LOG_LEVEL)
        .provide([[matchers.call.fn(sagas.getTimeStamp), FAKE_TIMESTAMP]])
        .run();

      expect(result.toJSON()).toMatchSnapshot();
    });
  });
});
