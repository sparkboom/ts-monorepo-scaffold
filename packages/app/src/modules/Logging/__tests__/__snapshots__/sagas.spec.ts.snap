// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logging > Sagas Log Saga should match the snapshot if content was successfully logged to browser 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {},
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/getTimeStamp",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "longUtc": "2019-10-4 20:39:58.328",
            "msSinceLaunch": "+45 ms",
            "shortLocal": "19-10-4 15:39:58.328",
          },
          "INFO",
          "LOG.CTX",
          "FAKE LOG MESSAGE",
          null,
          Array [
            Object {
              "details": "Fake Details",
            },
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "CALL",
    },
  ],
}
`;
